### ISSUE TITLE
Error with for-loop on symbols

### ISSUE THREAD

**Issue by georg-at-github** [apache/incubator-mxnet#5417]
Hi,

I am trying to implement an RNN in MXNet; as I have no GPU at my disposal, I sadly cannot use mxnet.symbol.rnn(). I use a for-loop for implementation and everything works fine, until exactly the 7th iteration ``(self.past_horizon+1==7)``. Then MXNet throws the following error message:

> [18:12:07] D:\Program Files (x86)\Jenkins\workspace\mxnet\mxnet\dmlc-core\include\dmlc/logging.h:300: [18:12:07] d:\program files (x86)\jenkins\workspace\mxnet\mxnet\src\operator\tensor\./matrix_op-inl.h:905: CropAssign only supports kWriteTo
[18:12:07] D:\Program Files (x86)\Jenkins\workspace\mxnet\mxnet\dmlc-core\include\dmlc/logging.h:300: [18:12:07] d:\program files (x86)\jenkins\workspace\mxnet\mxnet\src\engine\./threaded_engine.h:336: [18:12:07] d:\program files (x86)\jenkins\workspace\mxnet\mxnet\src\operator\tensor\./matrix_op-inl.h:905: CropAssign only supports kWriteTo
An fatal error occurred in asynchronous engine operation. If you do not know what caused this error, you can try set environment variable MXNET_ENGINE_TYPE to NaiveEngine and run with debugger (i.e. gdb). This will force all operations to be synchronous and backtrace will give you the series of calls that lead to this error. Remember to set MXNET_ENGINE_TYPE back to empty after debugging.

The error apparently occurs during the ``backward`` operation.

Here is my code:

    def __getgru__(self, data, num_hidden, name):
        hidden_state = mx.sym.zeros(shape=(self.batch_size, num_hidden, 1))
        ones = mx.sym.ones(shape=(self.batch_size, num_hidden, 1))
        W_z = mx.sym.Variable('W_z', shape=(self.batch_size, num_hidden, num_hidden), init=mx.init.Uniform(1))
        W_r = mx.sym.Variable('W_r', shape=(self.batch_size, num_hidden, num_hidden), init=mx.init.Uniform(1))
        W = mx.sym.Variable('W', shape=(self.batch_size, num_hidden, num_hidden), init=mx.init.Uniform(1))
        U_z = mx.sym.Variable('U_z', shape=(self.batch_size, num_hidden, num_hidden), init=mx.init.Uniform(1))
        U_r = mx.sym.Variable('U_r', shape=(self.batch_size, num_hidden, num_hidden), init=mx.init.Uniform(1))
        U = mx.sym.Variable('U', shape=(self.batch_size, num_hidden, num_hidden), init=mx.init.Uniform(1))
        outputs = mx.sym.Variable('outputs')
        data = mx.sym.swapaxes(data=data, dim1=1, dim2=2)
        for i in range(self.past_horizon+1):
            cur_data = mx.sym.slice(data=data, begin=(0, 0, i), end=(self.batch_size, num_hidden, i+1))
            reset_gate = mx.sym.Activation(
                mx.sym.batch_dot(W_r, cur_data) + mx.sym.batch_dot(U_r, hidden_state), act_type='sigmoid'
            )
            update_gate = mx.sym.Activation(
                mx.sym.batch_dot(W_z, cur_data) + mx.sym.batch_dot(U_z, hidden_state), act_type='sigmoid'
            )
            if i == 0:
                outputs = cur_data
                continue
            hidden_state = mx.sym.tanh(
                mx.sym.batch_dot(W, cur_data) + mx.sym.batch_dot(U, reset_gate * hidden_state)
            )
            update = (ones - update_gate) \
                     * mx.sym.slice(outputs, begin=(0, 0, i - 1), end=(self.batch_size, num_hidden, i)) \
                     + update_gate * hidden_state
            outputs = mx.sym.concat(outputs, update, dim=2)
        output = mx.sym.swapaxes(data=outputs, dim1=1, dim2=2)
        return(output)

I have tried out various types of input data, from which it is clear that this is not a memory issue. I can say with certainty that the problem lies with the ``__getgru__`` function, because the remaining NN works perfectly, when I skip the GRU. I have varied the other parameters (``num_hidden, batch_size``), and only ``past_horizon`` produces the error. Therefore my working hypothesis is that the computation of gradients through the loop only works for a certain number of steps.

Could you help me out here by suggesting a solution or an alternative?
Thank you so much!

**Comment by piiswrong** [comment]
try mx.rnn.GRUCell

**Comment by georg-at-github** [comment]
Thank you, I tried using ``mxnet.rnn.RNNCell`` instead (see code below). It runs smoothly, however after some 50 or so iterations, I start getting NaN's in the outputs/parameters. The data input definitely contains no missing values and are normalised to the interval [0,1]. Could you think of any explanations for this phenomenon?

I have noticed that the maximum value of ``fc2_weighted_output`` approaches 1 before NaN's begin occurring, i.e. the NN gets more specific (it only outputs one of two classes due to a highly imbalanced dataset). I cannot see why this should lead to NaN's being introduced, though.

        num_hidden = self.num_hidden
        bs = self.batch_size
        data = mx.sym.Variable('data', shape=(bs, self.past_horizon+1, self.num_vars))
        weights = mx.sym.Variable('weights', shape=(self.num_classes))
        weights = mx.sym.tile(weights, reps=(bs, 1))
        W_1 = mx.sym.Variable('W_1', shape=(num_hidden, self.num_vars), init=mx.init.Uniform(0.7))
        W_1 = mx.sym.tile(W_1, reps=(bs, 1, 1))
        b_1 = mx.sym.Variable('b_1', shape=(num_hidden), init=mx.init.Uniform(0.7))
        b_1 = mx.sym.tile(b_1, reps=(bs, self.past_horizon+1, 1))
        matmul = mx.sym.batch_dot(lhs=data, rhs=W_1, transpose_b=True, name='matmul')
        fc1 = matmul + b_1
        encoded = mx.sym.Activation(fc1, act_type='sigmoid', name='encoded')
        rnn = mx.rnn.RNNCell(num_hidden=num_hidden)
        outputs, states = rnn.unroll(length=self.past_horizon+1, inputs=encoded, merge_outputs=True)
        outputs = mx.sym.Reshape(outputs, shape=(bs,-1, num_hidden), name='ABC')
        outputs = mx.sym.max(outputs, axis=1, name='outputs')
        fc2 = mx.sym.FullyConnected(outputs, num_hidden=self.num_classes, name='fc2')
        fc2_weighted = mx.sym.add_n(fc2, weights, name='fc2_weighted')
        loss = mx.sym.SoftmaxOutput(fc2_weighted, name='softmax')

Again, thank you so much for any feedback!

**Comment by piiswrong** [comment]
try to tune learning rate initialization etc

**Comment by yajiedesign** [comment]
This issue is closed due to lack of activity in the last 90 days. Feel free to reopen if this is still an active issue. Thanks!
